### PR TITLE
[#150] Full JSON serialization for type preservation

### PR DESCRIPTION
**Pull request by delano**
## Summary

Implements complete JSON serialization for all Horreum field values to preserve Ruby types across the Redis storage boundary. Previously, some types were stored as raw strings (losing type information), while others used JSON inconsistently. Now all non-string values use JSON encoding for reliable type preservation.

## Key Changes

**Serialization Pipeline**
- All field values now use `JSON.dump` for consistent type preservation (Integer, Boolean, Float, nil, Hash, Array)
- Strings continue to be stored as-is to avoid double-quoting
- Encrypted fields automatically extract `encrypted_value` before storage
- Removed `dump_method`/`load_method` configuration options (hard-coded to JSON)

**Deserialization Updates**
- `deserialize_value` now returns all parsed JSON types, not just Hash/Array
- `find_by_dbkey` deserializes field values before object instantiation to ensure correct types
- Hash symbolization properly handles nested structures

**Bug Fixes**
- Fixed `initialize_with_keyword_args` to preserve `false` and `0` values (was using `||` operator which treated them as nil)
- Integer fields stay as Integer (35 → 35, not "35")
- Boolean fields preserve type (true → true, not "true")

**Encrypted Fields Integration**
- Updated `EncryptedData`, `Manager`, and `EncryptedFieldType` to accept both JSON strings (legacy) and pre-parsed Hashes (v2.0)
- Allows unified deserialization pipeline without special-case handling

**Module Refactoring**
- Renamed `Distinguisher` → `IdentifierExtractor` to clarify its actual purpose
- Role: Extract identifiers from Familia objects/classes for DataType storage (lists, sets, zsets)
- Retained simple type handling (String, Symbol, Integer, Float) for DataType primitive values
- Not used for general field serialization anymore (that's JSON's job now)

## Testing

Added comprehensive type preservation test suite covering:
- All primitive types: Integer, Boolean, String, Float, nil
- Complex types: Hash, Array with nested structures
- Round-trip consistency via save/load, batch_update, refresh!
- Hash symbolization edge cases

**All 2155 tests passing**

## Breaking Changes

⚠️ **Type Coercion Behavior Changed**

Fields that previously stored numbers as strings now preserve their Ruby types:

```ruby
# Before: stored as "35", loaded as "35" (String)
# After: stored as "35", loaded as 35 (Integer)
customer.age = 35
customer.save!
customer.refresh!
customer.age.class  # Integer (was String before)
```

**Migration**: If your code expects string-coerced numbers from Integer fields, update to handle actual Integer types.

## Documentation

- Updated CLAUDE.md with JSON serialization behavior
- Added changelog entry in `changelog.d/20251004_225209_delano_150_full_json_serialization.rst`
- Updated inline documentation for `serialize_value`, `deserialize_value`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

___

### **PR Type**
Enhancement

___

### **Description**
- Implement full JSON serialization for type preservation

- Fix falsy value handling in object initialization

- Rename Distinguisher to IdentifierExtractor for clarity

- Update encrypted field deserialization for unified pipeline

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Field Values"] --> B["serialize_value"]
  B --> C{"String?"}
  C -->|Yes| D["Store as-is"]
  C -->|No| E["JSON.dump"]
  D --> F["Redis Storage"]
  E --> F
  F --> G["deserialize_value"]
  G --> H{"JSON parseable?"}
  H -->|Yes| I["JSON.parse"]
  H -->|No| J["Return as string"]
  I --> K["Typed Value"]
  J --> K
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Miscellaneous</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>familia.rb</strong><dd><code>Update module require path for renamed file</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-2fcd125226b496da9c2e5992551dbd71828ea361f198c898bec403034fd97cf3">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>10 files</summary><table>
<tr>
  <td><strong>base.rb</strong><dd><code>Remove dump_method and load_method configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-adc0a2d820ca9b2723e3508e09df80ff20f72cc0a1daebeeb33936429de556a3">+0/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>serialization.rb</strong><dd><code>Update to use identifier_extractor instead of distinguisher</code></dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-94a4242b7f98ce39ccb6dc857726398f13d50fe8e868ad88a6bb0c667ebc0026">+5/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>distinguisher.rb</strong><dd><code>Remove entire file (renamed to identifier_extractor)</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-515f56a7213e1613b25cabfc1468b09b1f8393a2dfe4c76de71d6b1e6b8781ac">+0/-85</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>encrypted_data.rb</strong><dd><code>Support Hash input in from_json method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-552aa879730e826a0d4c1dd5ecca3762c2fec28dafcf95f8a66d295912ae07bd">+12/-2</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>manager.rb</strong><dd><code>Handle both JSON strings and Hashes in decrypt</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-f24fa737555abe9837940b0ead0019a9de9e406b503b7b9d65dee4c1fe5b2f18">+12/-4</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>encrypted_field_type.rb</strong><dd><code>Support Hash encrypted data and convert to JSON</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-40bceda392bd8906f94ad48b882415e55fadc2bd5e4636478791a74becc59604">+11/-3</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>definition.rb</strong><dd><code>Remove dump_method and load_method class methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-5d439fbd77f17e56a3c1ac01218164d4239a7a9f9b291cab396fceeb8e008174">+0/-12</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>management.rb</strong><dd><code>Add field deserialization in find_by_dbkey method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-555efc1b80ee6a419e468206f5de6642c5b2692708182b1fd2e5058d8e020b2e">+6/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>serialization.rb</strong><dd><code>Implement JSON serialization for all non-string values</code>&nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-e132748c31a61268f19736f1fd44db42ff4c3e3e8c811f36643c86e8ab7a4226">+29/-28</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>identifier_extractor.rb</strong><dd><code>New module for extracting identifiers from objects</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-c7dfb0b9c1020809667dcaa678ad0ae04a7c86b3a94ae28d880f6b5f9df51539">+60/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>horreum.rb</strong><dd><code>Fix falsy value handling in keyword args initialization</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-c42ea61e6354fd846082c12370a61f8e7771e39f1779388fad96d72658207300">+4/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td><strong>hash_symbolization_try.rb</strong><dd><code>Update tests for string keys default behavior</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-34344d2081bd51fed47bc53dd48220a8c7d3375c55f860e70dfae72fd063df8b">+30/-30</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>relationships_edge_cases_try.rb</strong><dd><code>Convert Time to Integer for JSON serialization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-808f5a1905c97b7a0a03886a572007349d8d97499059bda35d28ad524d5818d5">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>database_consistency_try.rb</strong><dd><code>Update expected values for type preservation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-432a0e34f4de9ccf2ffb0c68464d90ce894ad446c789e3cc938fa6c410fec323">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>relationships_persistence_round_trip_try.rb</strong><dd><code>Update tests for preserved Integer and nil types</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-4072e1d3b0d7e809f8ccdbe60c4df59cb4463724456608997b1ab40733ccbeb1">+16/-18</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>benchmarks_try.rb</strong><dd><code>Update benchmark to use JsonSerializer instead of distinguisher</code></dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-c1b7702f63039df4794797fe8c10b094c719aad998711b377a09707282c97130">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>base_enhancements_try.rb</strong><dd><code>Remove dump_method and load_method tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-564c2b1113cfc8d8d19484c219a71f8508cb17a7f03027a927fa87d7a7bc4a7e">+0/-8</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>utils_try.rb</strong><dd><code>Update tests for identifier_extractor functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-0809265f202586eecf61b2cf957c66c61235dd4f8dcd231142adcac2323e9dea">+17/-14</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>base_try.rb</strong><dd><code>Update test for Integer type preservation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-ca1d399dfd35e71466c0af04862caac8fe0137385d86f79e0200f92618164e4a">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>initialization_try.rb</strong><dd><code>Update test for Boolean type preservation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-c756b0f92d73e2f0c3f39444664ef48233c0270ce79286be7683875fd6a14deb">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>json_type_preservation_try.rb</strong><dd><code>Add comprehensive type preservation test suite</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-3161cea02ac4b30c7dd1466924c33558e522f6bb23746f63a1eaaf598ae5721b">+206/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>serialization_try.rb</strong><dd><code>Update tests for string keys in deserialization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-7ae8b4ea8092b33d199b9bb6b8f6c9b35532a3a383ead30bab32487b2a9d9a22">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>deserialization_correctness_test.rb</strong><dd><code>Add usage documentation comment</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-0a3ebdae922e9d55b4dcbed68e59888f6a3b9db18e0483cb0539efdda42e9eee">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>CLAUDE.md</strong><dd><code>Update documentation for JSON serialization behavior</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-6ebdb617a8104a7756d0cf36578ab01103dc9f07e4dc6feb751296b9c402faf7">+9/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>20251004_225209_delano_150_full_json_serialization.rst</strong><dd><code>Add changelog entry for breaking changes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-d83ee601d5beb85a183d2773e69cc6923b9ff7e70d06f130f9b893dfea6edab5">+39/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Formatting</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>pr_compliance_checklist.yaml</strong><dd><code>Fix YAML formatting consistency</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/delano/familia/pull/151/files#diff-58128063d894311f858a58a992b6fb22ba496063e2129ea753ecaf3523e3c30e">+6/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: lib/familia/horreum/serialization.rb [hunk 1]
```diff
@@ -177,20 +172,26 @@ def serialize_value(val)
       # @param symbolize [Boolean] Whether to symbolize hash keys (default: true)
       # @return [Object] The deserialized value (Hash, Array, or original string)
       #
```
[the remark this answers is not in the document]
Comment (delano) [reply lib/familia/horreum/serialization.rb:174]: This code hasn't been released yet

**Comment by delano** [comment]
@claude We've had a couple rounds of feedback and fixes.
